browser/ui/webui/skus_internals_ui.cc

```diff
@@ -294,7 +294,14 @@ void SkusInternalsUI::CreateOrderFromReceipt(
     CreateOrderFromReceiptCallback callback) {
   EnsureMojoConnected();
 
-  skus_service_->CreateOrderFromReceipt(domain, receipt, std::move(callback));
+  skus_service_->CreateOrderFromReceipt(
+      domain, receipt,
+      base::BindOnce(
+          [](CreateOrderFromReceiptCallback callback,
+             skus::mojom::SkusResultPtr result) {
+            std::move(callback).Run(result->message);
```

Comment:
Sorry, do you mind explaining what the additional layer of indirection does?

Reply:
What extra layer of indirection?
This is a mojom structure that contains a SkusResult on the heap, which came from Rust. You can't access the Rust version directly.

Reply:
I cannot pass `ffi::SkusResult` from Rust to `C++` directly as it contains a `rust::String` which is passed by reference even though the struct is pass by value. This causes a dangling pointer and crashes.

To solve it, I used `UniquePtr<ffi::SkusResult>::new` so that it's allocated on the heap and only deallocated when the c++ side is finished with it.

Because the structure is on the heap now, to access it requires `->`.
Since the structure can only be use at the point where it is given, the string is actually `copied` into an `std::string` in `skus::mojom::SkusResult` also on the heap, which is `skus::mojom::SkusResultPtr`.

So to access the string from c++ at a later time, that `SkusResultPtr` can be passed around to any functions, and then accessed without UB or deallocation or anything. To access the string requires `->` because it's a pointer.

Reply:
Thanks, that's helpful understanding the lifetimes involved.